File: Baekjoon/99mini/1987.py
'''
작성자: 99mini
문제 링크: https://www.acmicpc.net/problem/1987
카테고리
    - 그래프 이론
    - 그래프 탐색
    - 깊이 우선 탐색
    - 백트래킹
문제 해설
'''

import sys

input = sys.stdin.readline

r, c = map(int, input().split())

grid = [input() for _ in range(r)]

visited = [False for _ in range(ord('Z') - ord('A') + 1)]

dn = [
    [1,0],
    [-1,0],
    [0,1],
    [0,-1]
]

def is_boundary(row, col):
    return 0 <= row < r and 0 <= col < c

def solution():
    global ret
    ret = 0

    def dfs(row, col, depth):
        global ret
        ret = max(ret, depth)

        visited[ord(grid[row][col]) - ord('A')] = True

        for (dr, dc) in dn:
            next_row, next_col = row + dr, col + dc

            if not is_boundary(next_row, next_col):
                continue
            
            current_value = grid[next_row][next_col]

            if not visited[ord(current_value) - ord('A')]:
                dfs(next_row, next_col, depth + 1)
                visited[ord(current_value) - ord('A')] = False

    dfs(0, 0, 1)

    return ret

print(solution())

File: Baekjoon/99mini/1987.js
const readline = require("readline").createInterface({
  input: process.stdin,
  output: process.stdout,
});

const input = [];
readline
  .on("line", function (line) {
    input.push(line);
  })
  .on("close", function () {
    const [r, c] = input[0].split(" ");
    const grid = input.slice(1);

    let visited = Array.from({ length: "Z".charCodeAt() - "A".charCodeAt() + 1 }).fill(false);

    const dn = [
      [1, 0],
      [-1, 0],
      [0, 1],
      [0, -1],
    ];

    function isBoundary(row, col) {
      return 0 <= row && row < r && 0 <= col && col < c;
    }

    function solution() {
      let ret = 0;

      function dfs(row, col, depth) {
        ret = Math.max(ret, depth);

        visited[grid[row][col].charCodeAt() - "A".charCodeAt()] = true;

        for (const [dr, dc] of dn) {
          const [nextRow, nextCol] = [row + dr, col + dc];

          if (!isBoundary(nextRow, nextCol)) {
            continue;
          }

          const currentAlphaIndex = grid[nextRow][nextCol].charCodeAt() - "A".charCodeAt();

          if (!visited[currentAlphaIndex]) {
            dfs(nextRow, nextCol, depth + 1);
            visited[currentAlphaIndex] = false;
          }
        }
      }

      dfs(0, 0, 1);

      return ret;
    }

    console.log(solution());
    process.exit();
  });
